apis/v1beta1/config.go

```diff
@@ -474,23 +474,81 @@ func (s *Service) ApplyDefaults(logger logr.Logger) error {
 		return err
 	}
 
-	tm := &AnyConfig{
-		Object: map[string]interface{}{
-			"metrics": map[string]interface{}{
-				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
-			},
-		},
+	if s.Telemetry == nil {
+		s.Telemetry = &AnyConfig{
+			Object: map[string]any{},
+		}
 	}
 
-	if s.Telemetry == nil {
-		s.Telemetry = tm
+	metrics, ok := s.Telemetry.Object["metrics"]
+	if !ok {
+		metrics = map[string]any{}
+		s.Telemetry.Object["metrics"] = metrics
+	}
+	metricsMap, ok := metrics.(map[string]any)
+	if !ok {
+		logger.Info("metrics field is not a map")
+		return nil
+	}
+
+	readers, ok := metricsMap["readers"]
+	if !ok {
+		readers = []any{}
+		metricsMap["readers"] = readers
+	}
+	readersSlice, ok := readers.([]any)
+	if !ok {
+		logger.Info("readers field is not a slice")
 		return nil
 	}
-	// NOTE: Merge without overwrite. If a telemetry endpoint is specified, the defaulting
-	// respects the configuration and returns an equal value.
-	if err := mergo.Merge(s.Telemetry, tm); err != nil {
-		return fmt.Errorf("telemetry config merge failed: %w", err)
+
+	var found bool
+	for _, reader := range readersSlice {
```

Comment:
I've been preferring approaches where you use mapstructure to deserialize into a known struct instead of this map check / type assertion stuff. [example](https://github.com/open-telemetry/opentelemetry-operator/blob/17e463a5fd12ff52b7fd405fa734b8948584ef4a/internal/components/multi_endpoint.go#L37-L56)

I think we should do a similar approach for internal telemetry especially as setting will only get more advanced going forward.

Reply:
Thank you for your feedback! I'm not entirely sure I understand your suggestion, so let me check if I got it right.

Are you suggesting defining a complete struct for the Telemetry field and decoding it using mapstructure.Decode? If so, I personally feel that instead of decoding it with mapstructure, we should change the type of the Telemetry field from AnyConfig to the newly defined struct.

Or are you suggesting that we use the newly defined struct to read the internal telemetry configuration and set default values dynamically if they are missing? In that case, if I understand correctly, we would still need to perform type assertions like we do now, since we can't set default values without knowing the types.

Please let me know if I'm missing anything! 🙇

Reply:
I'm fine with this:
```
Are you suggesting defining a complete struct for the Telemetry field and decoding it using mapstructure.Decode? If so, I personally feel that instead of decoding it with mapstructure, we should change the type of the Telemetry field from AnyConfig to the newly defined struct.
```

Reply:
Initially we should stick with anyconfig, but we should be able to convert that anyconfig into the defined struct from go-contrib i.e. [this file](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/otelconf/v0.3.0/generated_config.go). If not, we can define a smaller intermediate type and convert to that. I want to avoid any `_, ok := ...` as much as we can.